### ISSUE TITLE
Query params are not set on relationship when reloading page

### ISSUE THREAD

**Issue by sarus** [mdehoog/ember-data-has-many-query#7]
I noticed in the ModelMixin that the relationship is not reloaded if it is not yet loaded.

```
    //if the hasMany content is already loaded, we must reload it
    if (value.get('content').get('isLoaded')) {
      value = value.reload();
    }
```

In my use case I have a route that is loading the model but on the very first load of the model the route wants to pass query params for when the hasMany relationship is fetched (the hasMany relationship looks like it is being called automatically when I display the relationship data in my template).

I was trying to do something like this in the `tag-entity-pair` model hook of my router:

```
    return this.store.findRecord('tag-entity-pair', params.tag_entity_pair_id, {adapterOptions: {query: params}}).then(function(tagEntityPair){
      // force the first fetch of the relationship to set the sort value of the relationship being fetched
      return tagEntityPair.query('comments', {sort: 'comment.last-edited'}).then(function(comments){
        return tagEntityPair;
      });
    });
```

However, because the `comments` have not been loaded at the time I run `tagEntityPair.query` the values are not reloaded and no query params are used.  

Is there a way to use the query params on the very first load of the relationship if it's being loaded from the model hook of the router?  

Apologies if this question makes no sense as I'm still trying to wrap my head around Ember Data (I have a feeling I might be setting up something wrong here).

Thank you for your effort on this plugin!

**Comment by mdehoog** [comment]
The reason for the `value.reload()` is, if the hasMany relationship data is already loaded, Ember Data will not reload from the backend again. If the relationship has not yet been loaded, the `query` method will automatically request data from the server, and therefore `reload` is not needed.

I suspect something else is going on. What does your `tag-entity-pair` model look like?

You could try returning the `tag-entity-pair` and the `comments` separately in the model hook (and change the template accordingly):

```javascript
    return this.store.findRecord('tag-entity-pair', params.tag_entity_pair_id, {adapterOptions: {query: params}}).then(function(tagEntityPair){
      // force the first fetch of the relationship to set the sort value of the relationship being fetched
      return tagEntityPair.query('comments', {sort: 'comment.last-edited'}).then(function(comments){
        return {
          tagEntityPair: tagEntityPair,
          comments: comments
        };
      });
    });
```

**Comment by sarus** [comment]
@mdehoog Thank you for the information on `value.reload()`.  That makes a lot of sense.  Here are the relevant parts of my two models:

```javascript
// models/tag-entity-pair.js
export default DS.Model.extend(HasManyQueryModelMixin, {
  comments: DS.hasMany('comment')
});
```

```javascript
// models/comment.js
export default DS.Model.extend(HasManyQueryModelMixin, {
  tagEntityPair: DS.belongsTo('tag-entity-pair'),
});
```

I added some output statements your two mixins and it seems like the following events are happening:
MODEL MIXIN: Enter
model.js:29 MODEL MIXIN: Set _commentsQueryParams to:
model.js:30 Object {tag_entity_pair_id: "6", sort: "-comment.last-edited", page: Object}
model.js:42 MODEL MIXIN: Setting _commentsQueryParams to undefined
1. Refresh the page
2. Model Mixin gets called
3. The model mixin is retrieving the query params correctly
```
{
   sort: comment.last-edited
}
```
4. The model mixin sets `_commentsQueryParams` properly
5. The content is not already loaded so we don't reload it
6.  value.finally returns and `_commentsQueryParams` is set to undefined
7. The Adapter mixin runs
8. The Adapter retrieves `_commentsQueryParam` and it is undefined
9.  The Adapter calls `this.ajax` with no query params since it was undefined

Based on the steps above, the ModelMixin appears to be setting the value of the comments query params to undefined before the AdapterMixing has a chance to run.

**Comment by mdehoog** [comment]
Thanks for the excellent analysis. Which version of Ember Data are you using? Does it help if you explicitly set `async: true` in your `DS.hasMany` options hash?

**Comment by sarus** [comment]
Here are the versions I'm running:

```
Ember             : 2.1.0
Ember Data        : 2.1.0
jQuery            : 1.11.3
```

Setting `{async:true}` explicitly doesn't seem to change anything.  The Adapter calls are still executed after the model mixin has set the query params back to `undefined`

**Comment by sarus** [comment]
Just updated to latest and still having the issue.  Will continue to play around with it and see if I can get any further with the debugging.

```
Ember             : 2.2.0
Ember Data        : 2.2.1
jQuery            : 1.11.3
```

**Comment by mdehoog** [comment]
I'm having trouble reproducing this issue. Any chance of a cut down example that demonstrates the issue?

The model mixin `query` function calls `get` for the passed hasMany attribute, which should trigger a the adapter to fetch the data, appending the query params to the URL. Here is my call stack:

![capture](https://cloud.githubusercontent.com/assets/1538523/11830129/3cd133a6-a3f6-11e5-92ea-cb84911088be.PNG)

You can see that the query function (currently selected) calls `this.get(propertyName);`, which ends up in this addon's `rest-adapter` mixin's `findHasMany` function. This adds the query parameters to the URL and calls ajax before the `value.finally` clears the query params. Perhaps you could set a breakpoint in the `rest-adapter` mixin `findHasMany` function and include your call stack here?

**Comment by sarus** [comment]
Thank you for investigating the issue.  I set breakpoints in both `model.js` where the `_queryParamProperty` gets set to null

```
//model.js
self.set(_queryParamPropertyName, undefined);
```

and then in the `rest-adapter` where the queryPramPropertyName is read back

```
//rest-adapter.js
var queryParams = snapshot.record.get(queryParamPropertyName(relationship.key));
```

and for whatever reason in my case the queryParam is set to undefined before the rest-adapter grabs the value.  

My call stack looks identical to yours.  I'm thinking it must be the way I'm trying to load the relationship in the model:

```
    return this.store.findRecord('tag-entity-pair', params.tag_entity_pair_id).then(function(tagEntityPair){
      return tagEntityPair.query('comments', params).then(function(comments) { //model.js query gets called here and the promise returns and sets the parameter to undefined
        return tagEntityPair; // this gets returned before any rest-adapter calls are made
      });
    });
```

I'll see if I can recreate it in a jsbin.

**Comment by octabode** [comment]
I think I've run into this same problem. Has anyone found a solution?

**Comment by sarus** [comment]
Sorry I'm still having the issue.  We've just lived with it for now until we have some cycles to look into it again:

We're on the following versions now:

```
Ember             : 2.3.1
Ember Data        : 2.3.3
jQuery            : 2.1.4
```

**Comment by mdehoog** [comment]
@sarus An update to the latest version of this addon and to ember-data >= 2.5 may help, I have switched the reloading logic to use the new ds-references implementation.

Note that this addon only works with an async HasMany relationship that is included in the parent model server response in the `links` attribute, e.g.
```javascript
{
    tag-entity-pair: {
        ...
        links: {comments: "comments"}
    }
}
```

**Comment by sarus** [comment]
Thank you for the info.  I'll get our app updated to `2.5` and report back.

**Comment by mdehoog** [comment]
I found a bug which is similar to this issue in that, when making two queries in quick succession, the first query is aborted and then query parameters are cleared, which causes them to be cleared for the second query too. I've fixed this issue in `v0.1.5` of this addon; it may help with this issue.
